src/compiler/lowering/MatMulOpLowering.cpp

```diff
@@ -14,198 +14,564 @@
+    unsigned NC = tile_sizes[4];
+    unsigned MC = tile_sizes[3];
+    unsigned KC = tile_sizes[2];
+    unsigned NR = tile_sizes[1];
+    unsigned MR = tile_sizes[0];
+    unsigned KU = options.unroll_factor;
+    auto vec_size = loops[1].getStep();
+    llvm::SmallVector<AffineForOp> loopNest;
+    getPerfectlyNestedLoops(loopNest, loops.front());
+    // tile i with MC, j with NC, k with KC
+    llvm::SmallVector<AffineForOp> tiledNest;
+    if (failed(tilePerfectlyNested(loopNest, {MC, NC, KC}, &tiledNest))) {
+      spdlog::warn("Could not tile the loop nest in MatMulLowering");
+    };
+    assert(tiledNest[0].getStep() == MC && "0 should have step size MC.");
+    assert(tiledNest[1].getStep() == NC * vec_size &&
+           "1 should have step size NC * vec_size.");
+    assert(tiledNest[2].getStep() == KC && "2 should have step size KC.");
+    assert(tiledNest[3].getStep() == 1 && "3 should have step size 1.");
+    assert(tiledNest[4].getStep() == 1 * vec_size &&
+           "4 should have step size vec_size.");
+    assert(tiledNest[5].getStep() == 1 && "5 should have step size 1.");
+
+    // Further tile the i mod MC loop with MR
+    if (failed(tilePerfectlyNested(tiledNest[3], {MR}))) {
+      spdlog::warn("Could not tile the second i loop in MatMulLowering");
+    };
+
+    // Further tile the j mod NC loop with NR
+    assert(tiledNest[4].getStep() == 1 * vec_size &&
+           "4 should have step size vec_size.");
+    if (failed(tilePerfectlyNested(tiledNest[4], {NR}))) {
+      spdlog::warn("Could not tile the second j loop in MatMulLowering");
+    };
+
+    llvm::SmallVector<AffineForOp> twiceTiledNest;
+    getPerfectlyNestedLoops(twiceTiledNest, tiledNest[0]);
+    assert(twiceTiledNest[0].getStep() == MC &&
+           "tTN: 0 should have step size MC."); // i loops
+    assert(twiceTiledNest[3].getStep() == MR &&
+           "tTN: 3 should have step size MR.");
+    assert(twiceTiledNest[4].getStep() == 1 &&
+           "tTN: 4 should have step size 1.");
+    assert(twiceTiledNest[1].getStep() == NC * vec_size &&
+           "tTN: 1 should have step size NC * vec_size."); // j loops
+    assert(twiceTiledNest[5].getStep() == NR * vec_size &&
+           "tTN: 5 should have step size NR.");
+    assert(twiceTiledNest[6].getStep() == 1 * vec_size &&
+           "tTN: 6 should have step size vec_size.");
+    assert(twiceTiledNest[2].getStep() == KC &&
+           "tTN: 2 should have step size 1."); // k loops
+    assert(twiceTiledNest[7].getStep() == 1 &&
+           "tTN: 7 should have step size 1.");
+
+    // permute loops to final order (i / MC, j / NC, k / KC, i / MR, i mod MR, j
+    // / NR, j mod NR, k mod KC) ->
+    //                              (j / NC, k / KC, i / MC, j / NR, i / MR, k
+    //                              mod KC, j mod NR, i mod MR)
+    // TODO: This assert only fails in debug mode?!
+    // assert(isValidLoopInterchangePermutation(twiceTiledNest, {2, 0, 1, 4, 7,
```

Comment:
Same as the `TODO` above.

Reply:
I think this check simply cannot deal with for loops that have more complex step lengths etc. Thus, I removed this check. The other checks should ensure we know where our loops are and therefore that this is a valid interchange.